scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/model/SchedulingActivity.java

```diff
@@ -0,0 +1,283 @@
+package gov.nasa.jpl.aerie.scheduler.model;
+
+
+import gov.nasa.jpl.aerie.constraints.model.EvaluationEnvironment;
+import gov.nasa.jpl.aerie.constraints.model.SimulationResults;
+import gov.nasa.jpl.aerie.constraints.time.Interval;
+import gov.nasa.jpl.aerie.constraints.tree.ProfileExpression;
+import gov.nasa.jpl.aerie.merlin.driver.ActivityDirective;
+import gov.nasa.jpl.aerie.merlin.driver.ActivityDirectiveId;
+import gov.nasa.jpl.aerie.merlin.protocol.types.Duration;
+import gov.nasa.jpl.aerie.merlin.protocol.types.SerializedValue;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * Contains all known information about an activity, representing BOTH directive-only information
+ * and instance-only information.
+ * If an activity only has a directive and hasn't been simulated, instance-specific data like duration
+ * will be null. If an activity is generated during simulation and thus doesn't have a directive, directive
+ * ID will be null. Only generated activities can have a non-null parent.
+ *
+ * @param id unique id
+ * @param type the descriptor for the behavior invoked by this activity instance
+ * @param startOffset the time at which this activity instance is scheduled to start
+ * @param duration the length of time this activity instances lasts for after its start
+ * @param arguments arguments are stored in a String/SerializedValue hashmap.
+ * @param topParent the parent activity if any
+ * @param isNew whether this activity was created in this scheduling run, or already existed in the plan
+ */
+public record SchedulingActivity(
+    ActivityDirectiveId id,
+    ActivityType type,
+    Duration startOffset,
+    Duration duration,
+    Map<String, SerializedValue> arguments,
+    ActivityDirectiveId topParent,
```

Comment:
Consider Optional here

Reply:
Yeah that's a valid concern. This class is generally a mess because of its dual meaning (it represents all known data about an activity, combining directive and/or instance). All of the fields except `type`, `startOffset` and `arguments` may or may not be valid depending on whether the activity has been simulated, is an instance-only spawned activity, has anchor, etc. The id fields may be `null`. The boolean fields will always be defined but their values may not mean anything depending on context.

I don't like it at all. But it is so deeply coupled to the rest of the scheduler that changing it, even to use `Optional`s, would be a pretty big ordeal I think.